polynote-kernel/src/main/scala/polynote/kernel/lang/scal/ScalaInterpreter.scala

```diff
@@ -38,37 +37,56 @@ class ScalaInterpreter(
 
   override def shutdown(): IO[Unit] = shutdownSignal.complete
 
-  protected val previousSources: mutable.HashMap[CellID, ScalaSource[this.type]] = new mutable.HashMap()
+  protected val previousSources: mutable.HashMap[CellID, ScalaSource[kernelContext.global.type]] = new mutable.HashMap()
 
-  protected lazy val notebookPackageName = global.TermName("$notebook")
-  def notebookPackage = global.Ident(notebookPackageName)
-  protected lazy val notebookPackageSymbol = global.internal.newModuleAndClassSymbol(global.rootMirror.RootPackage, notebookPackageName)
+  lazy val notebookPackageName = "$notebook"
+  lazy val notebookPackageTerm = global.TermName(notebookPackageName)
+  def notebookPackage = global.Ident(notebookPackageTerm)
+  protected lazy val notebookPackageSymbol = global.internal.newModuleAndClassSymbol(global.rootMirror.RootPackage, notebookPackageTerm)
 
-  // TODO: we want to get rid of predef and load `kernel` from the RST (whatever that ends up being)
+  // TODO: we want to get rid of predef and load `kernel` from the KernelContext
   def predefCode: Option[String] = Some("val kernel = polynote.runtime.Runtime")
 
   override def init(): IO[Unit] = IO.unit // pass for now
 
-  override def runCode(
-    cell: CellID,
-    visibleSymbols: Seq[Decl],
-    previousCells: Seq[CellID],
-    code: String
-  ): IO[Stream[IO, Result]] = {
-    val originalOut = System.out
+  // Compile and initialize the module, but don't reflect its values or output anything
+  def compileAndInit(source: ScalaSource[kernelContext.global.type], cellContext: CellContext): IO[Unit] = {
+    IO.fromEither(source.compile).flatMap {
+      case global.NoSymbol => IO.unit
+      case sym =>
+        val saveSource = IO.delay[Unit](previousSources.put(cellContext.id, source))
+        val setModule = cellContext.module.complete(sym.asModule)
+        setModule *> IO(kernelContext.runInterruptible(importToRuntime.importSymbol(sym))).map {
+          runtimeSym =>
+            kernelContext.runInterruptible {
+              val moduleMirror = try runtimeMirror.reflectModule(runtimeSym.asModule) catch {
+                case err: ExceptionInInitializerError => throw RuntimeError(err.getCause)
+                case err: Throwable =>
+                  throw RuntimeError(err) // could be linkage errors which won't get handled by IO#handleErrorWith
+              }
 
-    // TODO: this is wrong. It doesn't seem good to keep parsing this from "CellXX" though. Maybe cells also (or only?) need a numeric ID?
-    val cellIndex = previousCells.size
+              val instMirror = try runtimeMirror.reflect(moduleMirror.instance) catch {
+                case err: ExceptionInInitializerError => throw RuntimeError(err.getCause)
+                case err: Throwable =>
+                  throw RuntimeError(err)
+              }
+            }
+        } *> saveSource
+    }
+  }
```

Comment:
Any way to reuse `compileAndInit` inside `compileAndRun` or would that be too much of a bother?

Reply:
I think it would be more trouble than it's worth... they look similar but do slightly (importantly) different things, would be more confusing to try and break it up further IMHO